File: .github/hooks/scripts/orchestrated_user_prompt.py
from orchestrated_common import (
    load_hook_input,
    load_state,
    parse_feedback,
    print_json,
    repo_root_from_input,
    save_state,
    truncate_text,
    utc_now_iso,
)


def enqueue_task(state, prompt: str) -> str:
    state["taskCounter"] += 1
    task_name = truncate_text(prompt)
    task = {
        "id": f"task-{state['taskCounter']}",
        "name": task_name,
        "priority": 1,
        "payload": {"goal": prompt, "queuedAt": utc_now_iso()},
    }
    state["taskQueue"].append(task)
    state["activeTaskName"] = task_name
    return task_name


def main() -> int:
    input_data = load_hook_input()
    repo_root = repo_root_from_input(input_data)
    state = load_state(repo_root)
    prompt = input_data.get("prompt", "")
    normalized = prompt.strip().lower()

    if normalized == "resume":
        state["stopSignal"] = False
        state["conversationState"] = "idle"
        save_state(repo_root, state)
        return print_json({"systemMessage": "orchestrated-agent: stop mode cleared; execution may continue."})

    if normalized.startswith("prefer verbose"):
        state["preferences"]["verbosity"] = "high"
        save_state(repo_root, state)
        return print_json({"systemMessage": "orchestrated-agent: verbosity preference set to high."})

    if normalized.startswith("prefer brief") or normalized.startswith("prefer concise"):
        state["preferences"]["verbosity"] = "low"
        save_state(repo_root, state)
        return print_json({"systemMessage": "orchestrated-agent: verbosity preference set to low."})

    if normalized.startswith("prefer normal"):
        state["preferences"]["verbosity"] = "normal"
        save_state(repo_root, state)
        return print_json({"systemMessage": "orchestrated-agent: verbosity preference set to normal."})

    if normalized.startswith("prefer skill "):
        preferred_skill = prompt.strip()[len("prefer skill ") :].strip()
        state["preferences"]["preferredSkill"] = preferred_skill or None
        save_state(repo_root, state)
        return print_json({"systemMessage": f"orchestrated-agent: preferred skill hint set to '{preferred_skill}'."})

    feedback = parse_feedback(prompt)
    if feedback:
        state["feedbackLog"].append(
            {
                "task": state.get("activeTaskName") or "unspecified-task",
                "rating": feedback,
                "timestamp": utc_now_iso(),
                "skill": state["preferences"].get("preferredSkill"),
            }
        )
        if feedback == "Stop this request":
            state["stopSignal"] = True
            save_state(repo_root, state)
            return print_json({"systemMessage": "orchestrated-agent: stop mode enabled; future tool use will be denied until 'resume'."})

        save_state(repo_root, state)
        return print_json({"systemMessage": f"orchestrated-agent: recorded feedback '{feedback}' for '{state.get('activeTaskName') or 'unspecified-task'}'."})

    system_message = None
    conversation_state = state.get("conversationState", "idle")

    if conversation_state in {"idle", "awaitingGoal"}:
        task_name = enqueue_task(state, prompt)
        state["conversationState"] = "clarifying"
        system_message = (
            f"orchestrated-agent: queued goal '{task_name}'. "
            "Ask one concise clarifying question about constraints before editing or executing commands unless the prompt is already fully specified."
        )
    elif conversation_state == "clarifying":
        state["preferences"]["lastClarification"] = truncate_text(prompt, 240)
        state["conversationState"] = "executing"
        system_message = "orchestrated-agent: clarification captured; continue with the highest-priority queued task."
    else:
        task_name = enqueue_task(state, prompt)
        state["conversationState"] = "clarifying"
        system_message = (
            f"orchestrated-agent: current execution remains active and a follow-up goal '{task_name}' was queued. "
            "Finish the current task cleanly or explicitly switch context."
        )

    save_state(repo_root, state)
    return print_json({"systemMessage": system_message})


if __name__ == "__main__":
    raise SystemExit(main())

File: .github/hooks/scripts/orchestrated_user_prompt.mjs
import {
  detectRepoSkillHint,
  isStopRequest,
  loadHookInput,
  loadState,
  printJson,
  repoRootFromInput,
  runCli,
  saveState,
  truncateText,
} from './orchestrated_common.mjs';

function setActiveGoal(state, prompt) {
  const taskName = truncateText(prompt);
  state.activeTaskName = taskName;
  state.conversationState = 'executing';
  return taskName;
}

function parseGoalCommand(prompt) {
  const trimmed = prompt.trim();
  const match = /^(continue|clarify|switch|goal):\s*(.*)$/i.exec(trimmed);
  if (!match) {
    return null;
  }

  return {
    command: match[1].toLowerCase(),
    value: match[2].trim(),
  };
}

export async function main() {
  const inputData = await loadHookInput();
  const repoRoot = repoRootFromInput(inputData);
  const state = await loadState(repoRoot);
  const prompt = typeof inputData.prompt === 'string' ? inputData.prompt : '';
  const normalized = prompt.trim().toLowerCase();
  const goalCommand = parseGoalCommand(prompt);
  const repoSkillHint = detectRepoSkillHint(prompt);

  state.autoSkillHint = repoSkillHint?.key ?? null;

  if (normalized === 'resume') {
    state.stopSignal = false;
    state.conversationState = 'idle';
    state.activeTaskName = null;
    await saveState(repoRoot, state);
    return printJson({ systemMessage: 'orchestrated-agent: stop mode cleared; execution may continue.' });
  }

  if (normalized.startsWith('prefer verbose')) {
    state.preferences.verbosity = 'high';
    await saveState(repoRoot, state);
    return printJson({ systemMessage: 'orchestrated-agent: verbosity preference set to high.' });
  }

  if (normalized.startsWith('prefer brief') || normalized.startsWith('prefer concise')) {
    state.preferences.verbosity = 'low';
    await saveState(repoRoot, state);
    return printJson({ systemMessage: 'orchestrated-agent: verbosity preference set to low.' });
  }

  if (normalized.startsWith('prefer normal')) {
    state.preferences.verbosity = 'normal';
    await saveState(repoRoot, state);
    return printJson({ systemMessage: 'orchestrated-agent: verbosity preference set to normal.' });
  }

  if (normalized.startsWith('prefer skill ')) {
    const preferredSkill = prompt.trim().slice('prefer skill '.length).trim();
    state.preferences.preferredSkill = preferredSkill || null;
    await saveState(repoRoot, state);
    return printJson({
      systemMessage: `orchestrated-agent: preferred skill hint set to '${preferredSkill}'.`,
    });
  }

  if (isStopRequest(prompt)) {
    state.stopSignal = true;
    await saveState(repoRoot, state);
    return printJson({
      systemMessage:
        "orchestrated-agent: stop mode enabled; future tool use will be denied until 'resume'.",
    });
  }

  if (goalCommand) {
    if ((goalCommand.command === 'goal' || goalCommand.command === 'switch') && !goalCommand.value) {
      return printJson({
        systemMessage: `orchestrated-agent: '${goalCommand.command}:' requires a short goal after the colon.`,
      });
    }

    if ((goalCommand.command === 'continue' || goalCommand.command === 'clarify') && state.activeTaskName === null) {
      state.conversationState = 'awaitingGoal';
      await saveState(repoRoot, state);
      return printJson({
        systemMessage: `orchestrated-agent: no active goal is set. Start with 'goal: ...' before using '${goalCommand.command}:'.`,
      });
    }

    if (goalCommand.command === 'goal') {
      const taskName = setActiveGoal(state, goalCommand.value);
      await saveState(repoRoot, state);
      return printJson({
        systemMessage:
          `orchestrated-agent: active goal set to '${taskName}'. ` +
          'Treat later continue:/clarify: prompts as work on this same goal.',
      });
    }

    if (goalCommand.command === 'switch') {
      const taskName = setActiveGoal(state, goalCommand.value);
      await saveState(repoRoot, state);
      return printJson({
        systemMessage:
          `orchestrated-agent: switched active goal to '${taskName}'. ` +
          'Previous goal context is replaced.',
      });
    }

    state.conversationState = 'executing';
    await saveState(repoRoot, state);

    if (goalCommand.command === 'clarify') {
      const clarification = goalCommand.value || 'no extra clarification provided';
      return printJson({
        systemMessage:
          `orchestrated-agent: clarification recorded for active goal '${state.activeTaskName}': '${truncateText(clarification)}'. ` +
          'Keep working on the same goal and use this as added constraint/context.',
      });
    }

    const continuation = goalCommand.value || 'continue from the current state';
    return printJson({
      systemMessage:
        `orchestrated-agent: continue the active goal '${state.activeTaskName}' with '${truncateText(continuation)}'. ` +
        'Do not switch goals unless the user explicitly uses switch: or goal:.',
    });
  }

  let systemMessage;
  const { conversationState } = state;

  if (conversationState === 'idle' || conversationState === 'awaitingGoal' || state.activeTaskName === null) {
    const taskName = setActiveGoal(state, prompt);
    systemMessage =
      `orchestrated-agent: active goal set to '${taskName}'. ` +
      'Ask one concise clarifying question about constraints before editing or executing commands unless the prompt is already fully specified.';
  } else {
    const taskName = setActiveGoal(state, prompt);
    systemMessage =
      `orchestrated-agent: switched active goal to '${taskName}'. ` +
      'Ask one concise clarifying question before editing or executing commands unless the prompt is already fully specified.';
  }

  if (repoSkillHint) {
    systemMessage += ` ${repoSkillHint.instruction}`;
  }

  await saveState(repoRoot, state);
  return printJson({ systemMessage });
}

await runCli(main);